### PR TITLE
[core] Plumb preemption error through to user

### LINKED ISSUE #41107: [core] Node preemption information not surfaced upon actor death
When an actor dies due to its node being preempted, the surfaced error is indistinguishable from any other actor death. Partition the space of actor death errors more finely to surface preemption (retryable death) to the user.

### PR DESCRIPTION
**Pull request by vitsai**
## Why are these changes needed?

Currently, when a node is failed with a preemption error, the information is not propagated to actor failure message. This is the first PR to add preemption info to the actor failure

## Related issue number

Closes #41107

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: src/ray/gcs/gcs_server/gcs_node_manager.cc [hunk 1]
```diff
@@ -296,6 +301,29 @@ void GcsNodeManager::Initialize(const GcsInitData &gcs_init_data) {
          const std::pair<NodeID, int64_t> &right) { return left.second < right.second; });
 }
 
+std::optional<std::shared_ptr<rpc::GcsNodeInfo>> GcsNodeManager::GetDeadNode(
+    const ray::NodeID &node_id) const {
```
Comment (rkooo567) [inline src/ray/gcs/gcs_server/gcs_node_manager.cc:305]: we should not have blocking IO inside gcs

File: src/ray/gcs/gcs_server/gcs_node_manager.cc [hunk 2]
```diff
@@ -255,6 +259,7 @@ void GcsNodeManager::OnNodeFailure(const NodeID &node_id) {
   if (auto node = RemoveNode(node_id, /* is_intended = */ false)) {
     node->set_state(rpc::GcsNodeInfo::DEAD);
     node->set_end_time_ms(current_sys_time_ms());
+    node->mutable_death_info()->set_reason(rpc::NodeDeathInfo::UNEXPECTED_TERMINATION);
```
Comment (rkooo567) [inline src/ray/gcs/gcs_server/gcs_node_manager.cc:262]: we should pass the argument here?

File: src/ray/protobuf/common.proto [hunk 3]
```diff
@@ -343,6 +343,8 @@ message ActorDiedErrorContext {
   // Whether the actor had never started running before it died, i.e. it was cancelled
   // before scheduling had completed.
   bool never_started = 10;
+  // Whether the actor was on a preempted node.
+  bool preempted = 11;
```
Comment (rkooo567) [inline src/ray/protobuf/common.proto:347]: I think we can actually just insert node death protobuf here if it is due to node death?

**Comment by rkooo567** [comment]
can you push the latest changes?

**Comment by rkooo567** [comment]
Is this ready to be reviewed btw? Seems premerge is still failing

File: src/ray/gcs/gcs_server/gcs_node_manager.cc (hunk 2 above)
Comment (vitsai) [reply src/ray/gcs/gcs_server/gcs_node_manager.cc]: The detail has the autoscaler.DrainNodeReason though

File: src/ray/gcs/gcs_server/gcs_node_manager.cc [hunk 4]
```diff
@@ -282,6 +306,10 @@ void GcsNodeManager::OnNodeFailure(const NodeID &node_id,
   if (auto node = RemoveNode(node_id, /* is_intended = */ false)) {
     node->set_state(rpc::GcsNodeInfo::DEAD);
     node->set_end_time_ms(current_sys_time_ms());
+    if (node->death_info().reason() == rpc::NodeDeathInfo::UNSPECIFIED) {
```
Comment (rkooo567) [inline src/ray/gcs/gcs_server/gcs_node_manager.cc:309]: Add SetNodePreempted

and add unit test

`# SetNodePreempted -> OnNodeFailure -> verify reason is correctly set`

File: src/ray/gcs/gcs_server/gcs_autoscaler_state_manager.cc [hunk 5]
```diff
@@ -371,18 +374,38 @@ void GcsAutoscalerStateManager::HandleDrainNode(
     return;
   }
 
+  auto node = std::move(maybe_node.value());
```
Comment (rkooo567) [inline src/ray/gcs/gcs_server/gcs_autoscaler_state_manager.cc:377]: Add unit test to verify SetPreempted is called?

Comment (rkooo567) [reply src/ray/gcs/gcs_server/gcs_autoscaler_state_manager.cc:377]: ADd unit test to check publish happens

Comment (rkooo567) [reply src/ray/gcs/gcs_server/gcs_autoscaler_state_manager.cc:377]: up to you if you want to add it or not

File: src/ray/gcs/gcs_server/gcs_node_manager.cc [hunk 6]
```diff
@@ -282,6 +306,10 @@ void GcsNodeManager::OnNodeFailure(const NodeID &node_id,
...
+      // There was no drain in progress.
+      node->mutable_death_info()->set_reason(rpc::NodeDeathInfo::UNEXPECTED_TERMINATION);
```
[the remark this answers is not in the document]
Comment (rkooo567) [reply src/ray/gcs/gcs_server/gcs_node_manager.cc:311]: Let's do this in the next PR

File: src/ray/core_worker/transport/direct_actor_task_submitter.h [hunk 7]
```diff
@@ -366,6 +378,9 @@ class CoreWorkerDirectActorTaskSubmitter
   /// Used to complete tasks.
   TaskFinisherInterface &task_finisher_;
 
+  /// GCS client.
```
[the remark this answers is not in the document]
Comment (vitsai) [reply src/ray/core_worker/transport/direct_actor_task_submitter.h:393]: removed it

File: src/ray/gcs/gcs_server/gcs_actor_manager.cc [hunk 8]
```diff
@@ -210,18 +200,41 @@ void GcsActor::SetGrantOrReject(bool grant_or_reject) {
   grant_or_reject_ = grant_or_reject;
 }
 
+const ray::rpc::ActorDeathCause GcsActorManager::GenNodeDiedCause(
```
[the remark this answers is not in the document]
Comment (vitsai) [reply src/ray/gcs/gcs_server/gcs_actor_manager.cc:203]: keeping the function but removed GetDeadNode call

File: python/ray/tests/test_draining.py [hunk 9]
```diff
@@ -248,6 +248,46 @@ def ping(self):
     ray.get(obj, timeout=2) == head_node_id
 
 
+def test_draining_reason(ray_start_cluster):
```
[the remark this answers is not in the document]
Comment (rkooo567) [reply python/ray/tests/test_draining.py:251]: comment not addressed yet

File: src/ray/common/ray_config_def.h [hunk 10]
```diff
@@ -886,3 +886,7 @@ RAY_CONFIG(bool, enable_autoscaler_v2, false)
 // Python GCS client number of reconnection retry and timeout.
 RAY_CONFIG(int64_t, nums_py_gcs_reconnect_retry, 5)
 RAY_CONFIG(int64_t, py_gcs_connect_timeout_s, 30)
+
+// Whether to reap actor death reason from GCS.
+// Costs an extra RPC.
```
Comment (rkooo567) [inline src/ray/common/ray_config_def.h:891]: Add TODO to remove this?

File: src/ray/core_worker/transport/direct_actor_task_submitter.cc [hunk 11]
```diff
@@ -342,28 +342,56 @@ void CoreWorkerDirectActorTaskSubmitter::DisconnectActor(
...
+
 void CoreWorkerDirectActorTaskSubmitter::CheckTimeoutTasks() {
-  std::vector<std::pair<TaskSpecification, Status>> task_specs_and_status;
+  // The bool here indicates whether the task is the first for the raylet address.
+  auto task_info_list = std::make_shared<TaskInfoList>();
   {
     absl::MutexLock lock(&mu_);
     for (auto &queue_pair : client_queues_) {
       auto &queue = queue_pair.second;
+      if (!queue.rpc_client) {
```
Comment (rkooo567) [inline src/ray/core_worker/transport/direct_actor_task_submitter.cc:368]: do we need this logic? It didn't exist before

File: src/ray/core_worker/transport/direct_actor_task_submitter.h [hunk 12]
```diff
@@ -229,6 +240,10 @@ class CoreWorkerDirectActorTaskSubmitter
   void RetryCancelTask(TaskSpecification task_spec, bool recursive, int64_t milliseconds);
 
  private:
+  typedef std::vector<
```
Comment (rkooo567) [inline src/ray/core_worker/transport/direct_actor_task_submitter.h:243]: This is weird.. can we just create a struct or sth?

File: src/ray/core_worker/transport/direct_actor_task_submitter.h [hunk 13]
```diff
@@ -263,6 +278,8 @@ class CoreWorkerDirectActorTaskSubmitter
     /// indicate that the actor is not yet created. This is used to drop stale
     /// messages from the GCS.
     int64_t num_restarts = -1;
+    /// Whether this actor has received a preemption request.
```
Comment (rkooo567) [inline src/ray/core_worker/transport/direct_actor_task_submitter.h:281]:
```suggestion
    /// Whether this actor exits by spot preemption.
```

File: src/ray/gcs/gcs_server/gcs_actor_manager.cc [hunk 14]
```diff
@@ -1097,6 +1108,27 @@ void GcsActorManager::OnNodeDead(const NodeID &node_id,
   }
 }
 
+void GcsActorManager::SetPreemptedAndPublish(const NodeID &node_id) {
+  if (created_actors_.find(node_id) == created_actors_.end()) {
```
Comment (rkooo567) [inline src/ray/gcs/gcs_server/gcs_actor_manager.cc:1112]: comment the actor is already dead

File: src/ray/gcs/gcs_server/gcs_actor_manager.cc [hunk 15]
```diff
@@ -1097,6 +1108,27 @@ void GcsActorManager::OnNodeDead(const NodeID &node_id,
...
+    return;
+  }
+  for (const auto &id_iter : created_actors_.find(node_id)->second) {
```
Comment (rkooo567) [inline src/ray/gcs/gcs_server/gcs_actor_manager.cc:1115]: can you RAY_CHECK the state is not DEAD?

File: src/ray/gcs/gcs_server/gcs_actor_manager.cc [hunk 16]
```diff
@@ -1097,6 +1108,27 @@ void GcsActorManager::OnNodeDead(const NodeID &node_id,
...
+    auto actor_iter = registered_actors_.find(id_iter.second);
+    RAY_CHECK(actor_iter != registered_actors_.end());
```
Comment (rkooo567) [inline src/ray/gcs/gcs_server/gcs_actor_manager.cc:1117]: Add comment when assertion failed

File: src/ray/gcs/gcs_server/test/gcs_state_test.cc [hunk 17]
```diff
@@ -0,0 +1,184 @@
...
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// clang-format off
+#include <memory>
+
+#include "gmock/gmock.h"
```
Comment (rkooo567) [inline src/ray/gcs/gcs_server/test/gcs_state_test.cc:18]: let's remove this

**Review by rkooo567 — COMMENTED** [review]
https://github.com/ray-project/ray/pull/41102/files#r1402747573

this can be handled in the next PR. Let's address other comments and test spot failure e2e asap

File: python/ray/exceptions.py [hunk 18]
```diff
@@ -311,6 +313,9 @@ def __init__(self, cause: Union[RayTaskError, ActorDiedErrorContext] = None):
                 error_msg_lines.append(
                     "The actor never ran - it was cancelled before it started running."
                 )
+            if cause.preempted:
+                self.preempted = True
+                error_msg_lines.append("\tThe actor's node was preempted.")
```
Comment (jjyao) [inline python/ray/exceptions.py:318]: Are we going to show the actual preempt reason? Spot or fallback?

File: python/ray/exceptions.py [hunk 19]
```diff
@@ -326,6 +331,12 @@ def from_task_error(task_error: RayTaskError):
         return RayActorError(task_error)
 
 
+class RayPreemptionError(RayActorError):
```
Comment (jjyao) [inline python/ray/exceptions.py:334]: So we don't support task?

Comment (rkooo567) [reply python/ray/exceptions.py:334]: we are going to merge the minimal needed now for this PR (and follow up will contain other cases).

NOTE: Currently, task error architecture is not that great, and we cannot really detect correct error

File: src/ray/core_worker/task_manager.cc [hunk 20]
```diff
@@ -1021,6 +1021,7 @@ bool TaskManager::FailOrRetryPendingTask(const TaskID &task_id,
                                          const rpc::RayErrorInfo *ray_error_info,
                                          bool mark_task_object_failed,
                                          bool fail_immediately) {
+  // TODO(vitsai) do we even get the right error
```
[the remark this answers is not in the document]
Comment (vitsai) [reply src/ray/core_worker/task_manager.cc:1038]: removed it

File: python/ray/exceptions.py (hunk 18 above)
Comment (rkooo567) [reply python/ray/exceptions.py:318]: not in the scope of this PR. Is this reported as drain reason?

Comment (jjyao) [reply python/ray/exceptions.py:318]: Yes.

Comment (vitsai) [reply python/ray/exceptions.py:318]: This information is not currently reported from autoscaler.

File: BUILD.bazel [hunk 21]
```diff
@@ -1816,6 +1816,22 @@ ray_cc_test(
     ],
 )
 
+ray_cc_test(
+    name = "gcs_state_test",
```
Comment (rkooo567) [inline BUILD.bazel:1820]: remove it?

File: python/ray/exceptions.py [hunk 22]
```diff
@@ -280,6 +280,8 @@ def __init__(self, cause: Union[RayTaskError, ActorDiedErrorContext] = None):
         self._actor_init_failed = False
         # -- The base actor error message. --
         self.base_error_msg = "The actor died unexpectedly before finishing this task."
+        # Whether the node was preempted
+        self.preempted = False
```
Comment (rkooo567) [inline python/ray/exceptions.py:284]:
```suggestion
        self._preempted = False
```
And create a property to access it?

File: src/mock/ray/gcs/pubsub/gcs_pub_sub.h [hunk 23]
```diff
@@ -15,13 +15,20 @@
 namespace ray {
 namespace gcs {
 
-class MockGcsPubSub : public GcsPubSub {
+class MockGcsPublisher : public GcsPublisher {
  public:
+  MockGcsPublisher() {}
```
Comment (rkooo567) [inline src/mock/ray/gcs/pubsub/gcs_pub_sub.h:20]: Are these mock change necessary?

File: src/mock/ray/gcs/gcs_server/gcs_actor_manager.h [hunk 24]
```diff
@@ -27,45 +27,58 @@ namespace gcs {
 
 class MockGcsActorManager : public GcsActorManager {
  public:
+  MockGcsActorManager(GcsNodeManager &gcs_node_manager,
```
Comment (rkooo567) [inline src/mock/ray/gcs/gcs_server/gcs_actor_manager.h:30]: Remove all unrelated mock changes

File: src/ray/core_worker/transport/direct_actor_task_submitter.h [hunk 25]
```diff
@@ -61,6 +63,8 @@ class CoreWorkerDirectActorTaskSubmitterInterface {
 
   virtual void CheckTimeoutTasks() = 0;
 
+  virtual void SetPreempted(const ActorID &actor_id) = 0;
```
Comment (rkooo567) [inline src/ray/core_worker/transport/direct_actor_task_submitter.h:66]:
```suggestion
  /// Mark that the corresponding actor is preempted (e.g., spot preemption).
  /// If you call this API, when an actor is dead, it will include preempted = true to death cause. 
  virtual void SetPreempted(const ActorID &actor_id) = 0;
```

File: src/ray/core_worker/transport/direct_actor_task_submitter.h (hunk 12 above)
Comment (rkooo567) [reply src/ray/core_worker/transport/direct_actor_task_submitter.h:243]: still not addresed

File: src/ray/gcs/gcs_server/gcs_actor_manager.cc [hunk 26]
```diff
@@ -210,18 +200,39 @@ void GcsActor::SetGrantOrReject(bool grant_or_reject) {
...
+    std::shared_ptr<rpc::GcsNodeInfo> node) {
+  ray::rpc::ActorDeathCause death_cause;
+  auto actor_died_error_ctx = death_cause.mutable_actor_died_error_context();
+  AddActorInfo(actor, actor_died_error_ctx);
+  actor_died_error_ctx->set_error_message(
+      absl::StrCat("The actor is dead because its node has died. Node Id: ",
+                   NodeID::FromBinary(node->node_id()).Hex()));
+
+  if (auto death_info = node->death_info();
+      death_info.reason() == rpc::NodeDeathInfo::AUTOSCALER_DRAIN &&
```
Comment (rkooo567) [inline src/ray/gcs/gcs_server/gcs_actor_manager.cc:215]: Add TODO to publish this info along with preempted info

File: src/ray/gcs/gcs_server/gcs_actor_manager.cc [hunk 27]
```diff
@@ -1097,6 +1108,31 @@ void GcsActorManager::OnNodeDead(const NodeID &node_id,
   }
 }
 
+void GcsActorManager::SetPreemptedAndPublish(const NodeID &node_id) {
+  // The node is dead, so we need to mark all of its actors preempted.
```
Comment (rkooo567) [inline src/ray/gcs/gcs_server/gcs_actor_manager.cc:1112]: this comment seems wrong. 

```suggestion
  // Actors in the node is already dead. The actor death cause should've been already published, so we do nothing
```

File: src/ray/gcs/gcs_server/gcs_actor_manager.h [hunk 28]
```diff
@@ -336,6 +337,9 @@ class GcsActorManager : public rpc::ActorInfoHandler {
   Status RegisterActor(const rpc::RegisterActorRequest &request,
                        RegisterActorCallback success_callback);
 
+  /// Set actors on the node as preempted and publish the actor information.
```
Comment (rkooo567) [inline src/ray/gcs/gcs_server/gcs_actor_manager.h:340]:
```suggestion
  /// Set actors on the node as preempted and publish the actor information.
  /// If the node is already dead, it will be no-op.
```

File: src/ray/gcs/gcs_server/gcs_autoscaler_state_manager.cc [hunk 29]
```diff
@@ -371,18 +374,38 @@ void GcsAutoscalerStateManager::HandleDrainNode(
...
+
+  // Set the death reason of the node.
+  auto death_info = node->mutable_death_info();
```
Comment (rkooo567) [inline src/ray/gcs/gcs_server/gcs_autoscaler_state_manager.cc:380]: We should not directly access mutable field of node here. It is dangerous. 

Instead consider creating an node manager API to update the death info

File: src/ray/gcs/gcs_server/gcs_autoscaler_state_manager.cc [hunk 30]
```diff
@@ -371,18 +374,38 @@ void GcsAutoscalerStateManager::HandleDrainNode(
...
   raylet_client->DrainRaylet(
       request.reason(),
       request.reason_message(),
-      [reply, send_reply_callback](const Status &status,
-                                   const rpc::DrainRayletReply &raylet_reply) {
+      [this, reply, send_reply_callback, node_id](
+          const Status &status, const rpc::DrainRayletReply &raylet_reply) {
         reply->set_is_accepted(raylet_reply.is_accepted());
+
+        // Unset the death reason of the node if the drain was rejected.
```
Comment (rkooo567) [inline src/ray/gcs/gcs_server/gcs_autoscaler_state_manager.cc:400]: we need cpp unit tests for this.

File: src/ray/protobuf/BUILD [hunk 31]
```diff
@@ -29,6 +29,7 @@ proto_library(
     visibility = ["//java:__subpackages__"],
     deps = [
         ":common_proto",
+        ":autoscaler_proto",
```
Comment (rkooo567) [inline src/ray/protobuf/BUILD:32]: looks unnecessary?

**Review by rkooo567 — COMMENTED** [review]
Can you update the PR description? 

Also cc @jjyao can you review the PR?

File: src/ray/gcs/gcs_server/gcs_actor_manager.cc (hunk 27 above)
Comment (vitsai) [reply src/ray/gcs/gcs_server/gcs_actor_manager.cc:1112]: Yeah, it's wrong -- we set preempted before death and publish that state (when drain request arrives), then when it is actually dead we publish death.

File: src/ray/raylet/node_manager.cc [hunk 32]
```diff
@@ -1933,23 +1933,28 @@ void NodeManager::HandleDrainRaylet(rpc::DrainRayletRequest request,
...
-                 rpc::autoscaler::DrainNodeReason::DRAIN_NODE_REASON_PREEMPTION);
-    cluster_resource_scheduler_->GetLocalResourceManager().SetLocalNodeDraining();
+    cluster_resource_scheduler_->GetLocalResourceManager().SetLocalNodeDraining(
+        request.reason());
     reply->set_is_accepted(true);
+    break;
+  // Crash immediately.
+  case DrainNodeReason::DRAIN_NODE_REASON_UNSPECIFIED:
+  default:
+    RAY_LOG(ERROR) << "Drain raylet RPC has received, but the reason is not specified.";
```
Comment (jjyao) [inline src/ray/raylet/node_manager.cc:1957]: Why changing this? Before it will check fail but now it's only an error log

File: src/ray/raylet/scheduling/local_resource_manager.h [hunk 33]
```diff
@@ -221,6 +223,9 @@ class LocalResourceManager : public syncer::ReporterInterface {
 
   // Whether the local node is being drained or not.
   bool is_local_node_draining_ = false;
+  // Reason for drain, if draining.
+  DrainNodeReason drain_reason_ =
+      rpc::autoscaler::DrainNodeReason::DRAIN_NODE_REASON_UNSPECIFIED;
```
Comment (jjyao) [inline src/ray/raylet/scheduling/local_resource_manager.h:228]: Seems not used?

**Comment by rkooo567** [comment]
please ping me when comments are addressed

File: python/ray/exceptions.py [hunk 34]
```diff
@@ -280,6 +280,8 @@ def __init__(self, cause: Union[RayTaskError, ActorDiedErrorContext] = None):
         self._actor_init_failed = False
         # -- The base actor error message. --
         self.base_error_msg = "The actor died unexpectedly before finishing this task."
+        # Whether the node was preempted
+        self._preempted = False
```
Comment (rkooo567) [inline python/ray/exceptions.py:284]: Add

```python3
@property
def preempted
```

**Review by rkooo567 — APPROVED** [review]
@jjyao can you take the last look?

**Review by jjyao — APPROVED** [review]

File: src/ray/raylet/node_manager.cc (hunk 32 above)
Comment (jjyao) [reply src/ray/raylet/node_manager.cc]: Can we fix it?

Comment (rkooo567) [reply src/ray/raylet/node_manager.cc]: I fixed it

**Review by rkooo567 — COMMENTED** [review]
I removed drain_reason temporarily because it triggers java failure. It should be handled in a follow up PR. The follow up PR should

- cpp unit tests & do not modify the node state directly at autoscaler_state_manager.cc
- node failure info should be correctly set all the time
- proper exit code API
